File: services/backend/app/schemas/portfolio.py
"""
Portfolio Schemas
Pydantic models for portfolio data validation and serialization.
Following enterprise standards with proper financial data handling.
"""
from datetime import datetime, date
from decimal import Decimal
from typing import List, Optional, Dict, Any
from enum import Enum

from pydantic import BaseModel, Field, validator, root_validator
from pydantic.types import UUID4


class PositionType(str, Enum):
    """Position types following industry standards."""
    EQUITY = "equity"
    BOND = "bond"
    ETF = "etf"
    MUTUAL_FUND = "mutual_fund"
    OPTION = "option"
    FUTURE = "future"
    CRYPTO = "crypto"
    CASH = "cash"


class TransactionType(str, Enum):
    """Transaction types for audit compliance."""
    BUY = "buy"
    SELL = "sell"
    DIVIDEND = "dividend"
    SPLIT = "split"
    TRANSFER_IN = "transfer_in"
    TRANSFER_OUT = "transfer_out"
    FEE = "fee"
    INTEREST = "interest"


class PortfolioStatus(str, Enum):
    """Portfolio status following risk management standards."""
    ACTIVE = "active"
    INACTIVE = "inactive"
    RESTRICTED = "restricted"
    LIQUIDATING = "liquidating"
    CLOSED = "closed"


# Base schemas
class PortfolioPositionBase(BaseModel):
    """Base portfolio position schema."""
    symbol: str = Field(..., min_length=1, max_length=20, description="Stock symbol")
    position_type: PositionType = Field(..., description="Type of position")
    quantity: Decimal = Field(..., ge=0, description="Quantity held")
    average_cost: Decimal = Field(..., ge=0, description="Average cost per share")
    current_price: Optional[Decimal] = Field(None, ge=0, description="Current market price")
    last_updated: Optional[datetime] = Field(None, description="Last price update timestamp")
    
    @validator('quantity', 'average_cost', 'current_price', pre=True)
    def validate_decimal_precision(cls, v):
        """Ensure proper decimal precision for financial calculations."""
        if v is not None:
            return Decimal(str(v)).quantize(Decimal('0.01'))
        return v

    class Config:
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


class PortfolioPositionCreate(PortfolioPositionBase):
    """Schema for creating new portfolio positions."""
    pass


class PortfolioPositionUpdate(BaseModel):
    """Schema for updating portfolio positions."""
    quantity: Optional[Decimal] = Field(None, ge=0)
    average_cost: Optional[Decimal] = Field(None, ge=0)
    current_price: Optional[Decimal] = Field(None, ge=0)
    
    @validator('quantity', 'average_cost', 'current_price', pre=True)
    def validate_decimal_precision(cls, v):
        if v is not None:
            return Decimal(str(v)).quantize(Decimal('0.01'))
        return v


class PortfolioPosition(PortfolioPositionBase):
    """Complete portfolio position with calculated fields."""
    id: UUID4
    portfolio_id: UUID4
    market_value: Decimal = Field(..., description="Current market value")
    unrealized_pnl: Decimal = Field(..., description="Unrealized profit/loss")
    unrealized_pnl_percent: Decimal = Field(..., description="Unrealized P&L percentage")
    cost_basis: Decimal = Field(..., description="Total cost basis")
    weight: Decimal = Field(..., description="Portfolio weight percentage")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Portfolio schemas
class PortfolioBase(BaseModel):
    """Base portfolio schema."""
    name: str = Field(..., min_length=1, max_length=100, description="Portfolio name")
    description: Optional[str] = Field(None, max_length=500, description="Portfolio description")
    currency: str = Field(default="USD", min_length=3, max_length=3, description="Base currency")
    status: PortfolioStatus = Field(default=PortfolioStatus.ACTIVE, description="Portfolio status")


class PortfolioCreate(PortfolioBase):
    """Schema for creating new portfolios."""
    pass


class PortfolioUpdate(BaseModel):
    """Schema for updating portfolios."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=500)
    status: Optional[PortfolioStatus] = None


class Portfolio(PortfolioBase):
    """Complete portfolio schema with calculated metrics."""
    id: UUID4
    user_id: UUID4
    total_value: Decimal = Field(..., description="Total portfolio value")
    total_cost: Decimal = Field(..., description="Total cost basis")
    total_pnl: Decimal = Field(..., description="Total profit/loss")
    total_pnl_percent: Decimal = Field(..., description="Total P&L percentage")
    cash_balance: Decimal = Field(default=Decimal('0'), description="Cash balance")
    day_change: Decimal = Field(..., description="Today's change in value")
    day_change_percent: Decimal = Field(..., description="Today's change percentage")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Portfolio summary and analytics schemas
class PortfolioSummary(BaseModel):
    """Portfolio summary for dashboard display."""
    portfolio_id: UUID4
    portfolio_name: str
    total_value: Decimal
    day_change: Decimal
    day_change_percent: Decimal
    total_pnl: Decimal
    total_pnl_percent: Decimal
    cash_balance: Decimal
    positions_count: int
    last_updated: datetime
    
    class Config:
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


class PortfolioAnalytics(BaseModel):
    """Advanced portfolio analytics."""
    portfolio_id: UUID4
    total_value: Decimal
    asset_allocation: Dict[str, Decimal] = Field(..., description="Asset allocation by type")
    sector_allocation: Dict[str, Decimal] = Field(..., description="Sector allocation")
    top_holdings: List[Dict[str, Any]] = Field(..., description="Top 10 holdings")
    performance_metrics: Dict[str, Decimal] = Field(..., description="Performance metrics")
    risk_metrics: Dict[str, Decimal] = Field(..., description="Risk metrics")
    generated_at: datetime
    
    class Config:
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


class AIPortfolioInsight(BaseModel):
    """AI-generated portfolio insights."""
    portfolio_id: UUID4
    insight_type: str = Field(..., description="Type of insight (summary, recommendation, alert)")
    title: str = Field(..., max_length=200, description="Insight title")
    content: str = Field(..., description="Detailed insight content")
    confidence_score: float = Field(..., ge=0, le=1, description="AI confidence score")
    tags: List[str] = Field(default=[], description="Insight tags")
    action_required: bool = Field(default=False, description="Whether action is required")
    created_at: datetime
    expires_at: Optional[datetime] = Field(None, description="Insight expiration time")
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat() if v else None
        }


# Transaction schemas for audit compliance
class TransactionBase(BaseModel):
    """Base transaction schema for audit trails."""
    symbol: str = Field(..., min_length=1, max_length=20)
    transaction_type: TransactionType
    quantity: Decimal = Field(..., description="Transaction quantity")
    price: Decimal = Field(..., ge=0, description="Transaction price")
    fees: Decimal = Field(default=Decimal('0'), ge=0, description="Transaction fees")
    transaction_date: date = Field(..., description="Transaction date")
    notes: Optional[str] = Field(None, max_length=500, description="Transaction notes")
    
    @validator('quantity', 'price', 'fees', pre=True)
    def validate_decimal_precision(cls, v):
        if v is not None:
            return Decimal(str(v)).quantize(Decimal('0.01'))
        return v


class TransactionCreate(TransactionBase):
    """Schema for creating new transactions."""
    pass


class Transaction(TransactionBase):
    """Complete transaction with audit fields."""
    id: UUID4
    portfolio_id: UUID4
    user_id: UUID4
    total_amount: Decimal = Field(..., description="Total transaction amount")
    created_at: datetime
    created_by: UUID4
    
    class Config:
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Response schemas for API endpoints
class PortfolioResponse(BaseModel):
    """Standard portfolio response."""
    success: bool = True
    message: str
    data: Optional[Portfolio] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PortfolioDetailResponse(BaseModel):
    """Detailed portfolio response with positions and insights."""
    success: bool = True
    message: str
    data: Optional[Portfolio] = None
    positions: List[PortfolioPosition] = Field(default=[], description="Portfolio positions")
    insights: List[AIPortfolioInsight] = Field(default=[], description="AI insights")
    analytics: Optional[PortfolioAnalytics] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PortfolioPositionResponse(BaseModel):
    """Portfolio position response."""
    success: bool = True
    message: str
    data: Optional[PortfolioPosition] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class TransactionResponse(BaseModel):
    """Transaction response."""
    success: bool = True
    message: str
    data: Optional[Transaction] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class AIInsightResponse(BaseModel):
    """AI insight response."""
    success: bool = True
    message: str
    data: Optional[AIPortfolioInsight] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PortfolioListResponse(BaseModel):
    """Portfolio list response."""
    success: bool = True
    message: str
    data: List[Portfolio]
    total: int
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PortfolioSummaryResponse(BaseModel):
    """Portfolio summary response for dashboard."""
    success: bool = True
    message: str
    data: PortfolioSummary
    ai_insights: Optional[List[AIPortfolioInsight]] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PortfolioAnalyticsResponse(BaseModel):
    """Portfolio analytics response."""
    success: bool = True
    message: str
    data: PortfolioAnalytics
    timestamp: datetime = Field(default_factory=datetime.utcnow)


# Dashboard and performance schemas
class PerformanceMetrics(BaseModel):
    """Portfolio performance metrics that match frontend expectations."""
    total_return: float
    total_return_percentage: float
    day_return: float
    day_return_percentage: float
    week_return: float
    week_return_percentage: float
    month_return: float
    month_return_percentage: float
    year_return: float
    year_return_percentage: float
    sharpe_ratio: Optional[float] = None
    volatility: Optional[float] = None
    beta: Optional[float] = None
    max_drawdown: Optional[float] = None


class MarketSummary(BaseModel):
    """Market summary data that matches frontend expectations."""
    market_status: str  # 'OPEN' | 'CLOSED' | 'PRE_MARKET' | 'AFTER_HOURS'
    market_close_time: Optional[str] = None
    sp500_price: float
    sp500_change: float
    sp500_change_percentage: float
    nasdaq_price: float
    nasdaq_change: float
    nasdaq_change_percentage: float
    dow_price: float
    dow_change: float
    dow_change_percentage: float


class DashboardSummary(BaseModel):
    """Complete dashboard summary that matches frontend expectations."""
    portfolio: Portfolio
    positions: List[PortfolioPosition]
    recent_transactions: List[Transaction]
    ai_insights: List[AIPortfolioInsight]
    market_summary: MarketSummary
    performance_metrics: PerformanceMetrics
    
    class Config:
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        } 

File: src/types/portfolio.ts
/**
 * Portfolio Types
 * Type definitions for portfolio-related data structures
 * Aligned with backend Pydantic schemas for consistency
 */

// Base portfolio information
export interface Portfolio {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  total_value: number;
  cash_balance: number;
  total_cost: number;
  total_pnl: number;
  total_pnl_percentage: number;
  day_pnl: number;
  day_pnl_percentage: number;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

// Individual portfolio position
export interface PortfolioPosition {
  id: string;
  portfolio_id: string;
  symbol: string;
  quantity: number;
  average_cost: number;
  current_price: number;
  market_value: number;
  total_cost: number;
  unrealized_pnl: number;
  unrealized_pnl_percentage: number;
  day_pnl: number;
  day_pnl_percentage: number;
  weight_percentage: number;
  created_at: string;
  updated_at: string;
}

// Transaction record
export interface Transaction {
  id: string;
  portfolio_id: string;
  symbol: string;
  transaction_type: 'BUY' | 'SELL' | 'DIVIDEND' | 'SPLIT' | 'TRANSFER';
  quantity: number;
  price: number;
  total_amount: number;
  fees: number;
  transaction_date: string;
  created_at: string;
  notes?: string;
}

// AI-generated portfolio insights
export interface AIPortfolioInsight {
  id: string;
  portfolio_id: string;
  insight_type: 'ANALYSIS' | 'RECOMMENDATION' | 'ALERT' | 'SUMMARY';
  title: string;
  content: string;
  confidence_score: number;
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  created_at: string;
  expires_at?: string;
  is_read: boolean;
  metadata?: Record<string, any>;
}

// Portfolio snapshot for historical tracking
export interface PortfolioSnapshot {
  id: string;
  portfolio_id: string;
  snapshot_date: string;
  total_value: number;
  cash_balance: number;
  total_cost: number;
  total_pnl: number;
  total_pnl_percentage: number;
  positions_count: number;
  metadata?: Record<string, any>;
}

// Dashboard summary response
export interface DashboardSummary {
  portfolio: Portfolio;
  positions: PortfolioPosition[];
  recent_transactions: Transaction[];
  ai_insights: AIPortfolioInsight[];
  market_summary: MarketSummary;
  performance_metrics: PerformanceMetrics;
}

// Market summary data
export interface MarketSummary {
  market_status: 'OPEN' | 'CLOSED' | 'PRE_MARKET' | 'AFTER_HOURS';
  market_close_time?: string;
  sp500_price: number;
  sp500_change: number;
  sp500_change_percentage: number;
  nasdaq_price: number;
  nasdaq_change: number;
  nasdaq_change_percentage: number;
  dow_price: number;
  dow_change: number;
  dow_change_percentage: number;
}

// Performance metrics
export interface PerformanceMetrics {
  total_return: number;
  total_return_percentage: number;
  day_return: number;
  day_return_percentage: number;
  week_return: number;
  week_return_percentage: number;
  month_return: number;
  month_return_percentage: number;
  year_return: number;
  year_return_percentage: number;
  sharpe_ratio?: number;
  volatility?: number;
  beta?: number;
  max_drawdown?: number;
}

// API request types
export interface CreatePortfolioRequest {
  name: string;
  description?: string;
  initial_cash?: number;
}

export interface UpdatePortfolioRequest {
  name?: string;
  description?: string;
  cash_balance?: number;
}

export interface AddPositionRequest {
  symbol: string;
  quantity: number;
  price: number;
  transaction_type: 'BUY' | 'SELL';
  notes?: string;
}

export interface UpdatePositionRequest {
  quantity?: number;
  notes?: string;
}

// API response types
export interface PortfolioListResponse {
  portfolios: Portfolio[];
  total_count: number;
}

export interface PortfolioDetailResponse {
  portfolio: Portfolio;
  positions: PortfolioPosition[];
  recent_transactions: Transaction[];
  performance_metrics: PerformanceMetrics;
}

// Error types
export interface PortfolioError {
  code: string;
  message: string;
  details?: Record<string, any>;
}

// Loading states for UI
export interface PortfolioLoadingState {
  isLoading: boolean;
  isRefreshing: boolean;
  lastUpdated?: string;
  error?: PortfolioError;
}

// Chart data types
export interface ChartDataPoint {
  date: string;
  value: number;
  change?: number;
  change_percentage?: number;
}

export interface PortfolioChartData {
  labels: string[];
  values: number[];
  changes: number[];
  timeframe: 'DAY' | 'WEEK' | 'MONTH' | 'QUARTER' | 'YEAR' | 'ALL';
}

// Widget data for dashboard cards
export interface DashboardWidget {
  id: string;
  title: string;
  value: number | string;
  change?: number;
  change_percentage?: number;
  trend: 'UP' | 'DOWN' | 'NEUTRAL';
  subtitle?: string;
  icon?: string;
  color?: 'green' | 'red' | 'blue' | 'gray';
}

// Filter and sorting options
export interface PortfolioFilters {
  timeframe?: 'DAY' | 'WEEK' | 'MONTH' | 'QUARTER' | 'YEAR';
  asset_type?: 'STOCK' | 'ETF' | 'CRYPTO' | 'BOND' | 'OPTION';
  sort_by?: 'symbol' | 'value' | 'pnl' | 'percentage' | 'weight';
  sort_order?: 'ASC' | 'DESC';
  search?: string;
}

// Real-time updates
export interface PortfolioUpdate {
  type: 'PRICE_UPDATE' | 'POSITION_CHANGE' | 'TRANSACTION' | 'INSIGHT';
  portfolio_id: string;
  data: any;
  timestamp: string;
}

// Enhanced portfolio types for PortfolioPage
export interface Position {
  id: string;
  symbol: string;
  companyName: string;
  shares: number;
  avgCost: number;
  currentPrice: number;
  marketValue: number;
  gainLoss: number;
  gainLossPercentage: number;
  dayChange?: number;
  dayChangePercentage?: number;
  weight?: number;
  sector?: string;
  lastUpdated?: string;
}

export interface PortfolioSummary {
  totalValue: number;
  dayChange: number;
  dayChangePercentage: number;
  totalGainLoss: number;
  totalGainLossPercentage: number;
  openPositions: number;
  profitablePositions: number;
  cashBalance: number;
  cashPercentage: number;
  portfolioReturn: number;
  totalInvested: number;
}

export interface Holdings {
  positions: Position[];
  totalValue: number;
  lastUpdated: string;
}

export type HoldingsSortBy =
  | 'symbol'
  | 'companyName'
  | 'shares'
  | 'avgCost'
  | 'currentPrice'
  | 'marketValue'
  | 'gainLoss'
  | 'gainLossPercentage'
  | 'dayChange'
  | 'weight';

export type SortDirection = 'asc' | 'desc';

export interface HoldingsFilter {
  searchTerm: string;
  sector?: string;
  profitableOnly: boolean;
  minValue?: number;
  maxValue?: number;
  sortBy: HoldingsSortBy;
  sortDirection: SortDirection;
}

export interface PortfolioAnalytics {
  performanceMetrics: PerformanceMetrics;
  riskMetrics: RiskMetrics;
  diversificationMetrics: DiversificationMetrics;
  comparisonData: BenchmarkComparison;
}

export interface RiskMetrics {
  volatility: number;
  beta: number;
  sharpeRatio: number;
  maxDrawdown: number;
  valueAtRisk: number;
  conditionalValueAtRisk: number;
  correlationMatrix?: Record<string, Record<string, number>>;
}

export interface DiversificationMetrics {
  sectorAllocation: Array<{
    sector: string;
    percentage: number;
    value: number;
  }>;
  geographicAllocation: Array<{
    region: string;
    percentage: number;
    value: number;
  }>;
  marketCapAllocation: Array<{
    category: string;
    percentage: number;
    value: number;
  }>;
  concentrationRisk: number;
  herfindahlIndex: number;
}

export interface BenchmarkComparison {
  benchmark: string;
  portfolioReturn: number;
  benchmarkReturn: number;
  alpha: number;
  beta: number;
  correlationCoefficient: number;
  trackingError: number;
  informationRatio: number;
}

export interface PriceUpdate {
  symbol: string;
  currentPrice: number;
  change: number;
  changePercentage: number;
  timestamp: string;
  volume?: number;
  bid?: number;
  ask?: number;
}

export interface RealtimeData {
  prices: Record<string, PriceUpdate>;
  marketStatus: 'OPEN' | 'CLOSED' | 'PRE_MARKET' | 'AFTER_HOURS';
  lastUpdated: string;
}

export interface ExportOptions {
  format: 'CSV' | 'PDF' | 'EXCEL';
  includeTransactions: boolean;
  includeAnalytics: boolean;
  dateRange?: {
    start: string;
    end: string;
  };
  customFields?: string[];
}

export interface PositionAction {
  type:
    | 'BUY_MORE'
    | 'SELL'
    | 'SET_ALERT'
    | 'VIEW_DETAILS'
    | 'VIEW_NEWS'
    | 'ANALYZE';
  symbol: string;
  data?: any;
}

export interface StockNews {
  id: string;
  symbol: string;
  headline: string;
  summary: string;
  source: string;
  publishedAt: string;
  sentiment: 'POSITIVE' | 'NEGATIVE' | 'NEUTRAL';
  url: string;
}

export interface PriceAlert {
  id: string;
  symbol: string;
  alertType: 'PRICE_ABOVE' | 'PRICE_BELOW' | 'PERCENT_CHANGE' | 'VOLUME';
  threshold: number;
  isActive: boolean;
  createdAt: string;
  triggeredAt?: string;
}

export interface TableState {
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
  sortBy?: HoldingsSortBy;
  sortDirection?: SortDirection;
  filters: HoldingsFilter;
}

export interface PortfolioPageConfig {
  showSummaryCards: boolean;
  showAnalytics: boolean;
  showNews: boolean;
  showAlerts: boolean;
  enableRealTimeUpdates: boolean;
  refreshInterval: number;
  defaultPageSize: number;
  enableExport: boolean;
  enableQuickActions: boolean;
  compactMode: boolean;
}
